File: src/sdk/langchain.ts
// LangChain / LangGraph (JS) adapter: a callback handler that issues a receipt for every tool run it observes.
// Observe-only. LangChain callbacks cannot block a tool, so this handler evaluates no policy. For enforcement,
// construct the tool with issuer.wrap(): tool(issuer.wrap("name", fn), { name, schema }). Do not combine both on
// one tool, or it will be recorded twice.
import { BaseCallbackHandler } from "@langchain/core/callbacks/base";
import type { SdkIssuer, ToolEvent } from "./index.ts";

function parseArgs(input: string): Record<string, unknown> {
  try {
    const v = JSON.parse(input) as unknown;
    return v && typeof v === "object" && !Array.isArray(v) ? (v as Record<string, unknown>) : { input: v };
  } catch {
    return { input };
  }
}

/** LangChain hands the handler either the raw return value or a ToolMessage. Record the content either way. */
function unwrapOutput(output: unknown): unknown {
  if (output && typeof output === "object" && "content" in output && "tool_call_id" in output) {
    const content = (output as { content: unknown }).content;
    if (typeof content === "string") {
      try {
        return JSON.parse(content) as unknown;
      } catch {
        return content;
      }
    }
    return content;
  }
  return output;
}

export class ReceiptCallbackHandler extends BaseCallbackHandler {
  name = "agent-receipts";
  private readonly issuer: SdkIssuer;
  private readonly pending = new Map<string, ToolEvent>();

  constructor(issuer: SdkIssuer) {
    super();
    this.issuer = issuer;
  }

  override handleToolStart(
    tool: { name?: string },
    input: string,
    runId: string,
    _parentRunId?: string,
    _tags?: string[],
    _metadata?: Record<string, unknown>,
    runName?: string,
    toolCallId?: string,
  ): void {
    const name = runName ?? tool.name ?? "unknown";
    this.pending.set(runId, { tool: name, args: parseArgs(input), session: { id: null, toolUseId: toolCallId ?? null } });
  }

  override handleToolEnd(output: unknown, runId: string): void {
    const ev = this.pending.get(runId);
    if (!ev) return;
    this.pending.delete(runId);
    this.issuer.record(ev, { status: "executed", result: unwrapOutput(output) }, null);
  }

  override handleToolError(err: Error, runId: string): void {
    const ev = this.pending.get(runId);
    if (!ev) return;
    this.pending.delete(runId);
    this.issuer.record(ev, { status: "error", error: err.message }, null);
  }
}

/** Convenience: `tool.invoke(args, receiptCallbacks(issuer))`, or spread into any RunnableConfig. */
export function receiptCallbacks(issuer: SdkIssuer): { callbacks: ReceiptCallbackHandler[] } {
  return { callbacks: [new ReceiptCallbackHandler(issuer)] };
}

File: packages/python/agent_custody/langchain.py
"""LangChain / LangGraph adapter: a callback handler that records a receipt for every tool run it sees.

Observe-only. Callbacks cannot block a tool, so this evaluates no policy. For enforcement wrap the function before
turning it into a tool: `tool(client.wrap("name", fn))`. Do not combine both on one tool, or it is recorded twice.
"""
from __future__ import annotations

import json
from typing import Any, Dict, Optional
from uuid import UUID

from langchain_core.callbacks import BaseCallbackHandler

from . import Client


def _parse_args(input_str: str, inputs: Optional[Dict[str, Any]]) -> Dict[str, Any]:
    if isinstance(inputs, dict):
        return inputs
    try:
        v = json.loads(input_str)
        return v if isinstance(v, dict) else {"input": v}
    except (TypeError, ValueError):
        return {"input": input_str}


def _unwrap(output: Any) -> Any:
    content = getattr(output, "content", None) if hasattr(output, "tool_call_id") else None
    if content is None:
        return output
    if isinstance(content, str):
        try:
            return json.loads(content)
        except ValueError:
            return content
    return content


class ReceiptCallbackHandler(BaseCallbackHandler):
    name = "agent-custody"

    def __init__(self, client: Client):
        super().__init__()
        self.client = client
        self._pending: Dict[UUID, Dict[str, Any]] = {}

    def on_tool_start(self, serialized: Dict[str, Any], input_str: str, *, run_id: UUID, parent_run_id: Optional[UUID] = None, tags=None, metadata=None, inputs: Optional[Dict[str, Any]] = None, **kwargs: Any) -> None:
        name = kwargs.get("name") or (serialized or {}).get("name") or "unknown"
        self._pending[run_id] = {"tool": name, "args": _parse_args(input_str, inputs), "session": {"id": None, "toolUseId": kwargs.get("tool_call_id")}}

    def on_tool_end(self, output: Any, *, run_id: UUID, **kwargs: Any) -> None:
        ev = self._pending.pop(run_id, None)
        if ev is None:
            return
        self.client.record(ev["tool"], ev["args"], {"status": "executed", "result": _unwrap(output)}, None, session=ev["session"])

    def on_tool_error(self, error: BaseException, *, run_id: UUID, **kwargs: Any) -> None:
        ev = self._pending.pop(run_id, None)
        if ev is None:
            return
        self.client.record(ev["tool"], ev["args"], {"status": "error", "error": str(error)}, None, session=ev["session"])
